packages/world/ts/actions/callFrom.ts

```diff
@@ -0,0 +1,148 @@
+    writeContract: async (writeArgs): Promise<WriteContractReturnType> => {
+      // Skip if the contract isn't the World.
+      if (writeArgs.address !== params.worldAddress) {
+        return getAction(client, writeContract, "writeContract")(writeArgs);
+      }
+
+      // Encode the World's calldata (which includes the World's function selector).
+      const worldCalldata = encodeFunctionData({
+        abi: writeArgs.abi,
+        functionName: writeArgs.functionName,
+        args: writeArgs.args,
+      } as unknown as EncodeFunctionDataParameters);
+
+      // The first 4 bytes of calldata represent the function selector.
+      const worldFunctionSelector = slice(worldCalldata, 0, 4);
+
+      // Get the systemId and System's function selector.
+      const { systemId, systemFunctionSelector } = await worldFunctionToSystemFunction(params, worldFunctionSelector);
+
+      // Construct the System's calldata.
+      // If there's no args, use the System's function selector as calldata.
+      // Otherwise, use the World's calldata, replacing the World's function selector with the System's.
+      const systemCalldata =
+        worldCalldata === worldFunctionSelector
+          ? systemFunctionSelector
+          : concat([systemFunctionSelector, slice(worldCalldata, 4)]);
+
+      // Construct args for `callFrom`.
+      const callFromArgs: typeof writeArgs = {
+        ...writeArgs,
+        functionName: "callFrom",
+        args: [params.delegatorAddress, systemId, systemCalldata],
+      };
+
+      // Call `writeContract` with the new args.
+      return getAction(client, writeContract, "writeContract")(callFromArgs);
+    },
+  });
+}
+
+const systemFunctionCache = new Map<Hex, SystemFunction>();
+
+async function worldFunctionToSystemFunction(
+  params: CallFromParameters,
+  worldFunctionSelector: Hex,
+): Promise<SystemFunction> {
+  const cacheKey = concat([params.worldAddress, worldFunctionSelector]);
+
+  // Use cache if the function has been called previously.
+  const cached = systemFunctionCache.get(cacheKey);
+  if (cached) return cached;
+
+  // If a mapping function is provided, use it. Otherwise, call the World contract.
+  const systemFunction = params.worldFunctionToSystemFunction
+    ? await params.worldFunctionToSystemFunction(worldFunctionSelector)
+    : await retrieveSystemFunctionFromContract(params.publicClient, params.worldAddress, worldFunctionSelector);
+
+  systemFunctionCache.set(cacheKey, systemFunction);
+
+  return systemFunction;
+}
+
+async function retrieveSystemFunctionFromContract(
+  publicClient: PublicClient,
+  worldAddress: Hex,
+  worldFunctionSelector: Hex,
+): Promise<SystemFunction> {
+  const table = worldConfig.tables.world__FunctionSelectors;
+
+  const _keySchema = getKeySchema(table);
+  const keySchema = mapObject<typeof _keySchema, { [K in keyof typeof _keySchema]: (typeof _keySchema)[K]["type"] }>(
+    _keySchema,
+    ({ type }) => type,
+  );
```

Comment:
This `keySchema`'s type is `{ worldFunctionSelector: "bytes4" }`. Without specifying the generics type argument (using `mapObject(_keySchema, ...)`), the type results in `{ worldFunctionSelector: unknown }`. There might be a better way to manage this with the new config.

Reply:
ah, `getSchemaTypes`